cloud_storage/cloud_storage/overrides/file.py

```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
 				_("This file is attached to a submitted document and cannot be deleted"),
 				frappe.PermissionError,
 			)
-
+		if len(self.file_association) > 1:
+			frappe.throw(
+				_("This file is attached to multiple documents and cannot be deleted"),
+				frappe.PermissionError,
+			)
 		super().on_trash()
 
+	def associate_files(self) -> File:
+		if not self.content_hash and "/api/method/retrieve" in self.file_url:  # type: ignore
+			associated_doc = frappe.get_value("File", {"file_url": self.file_url}, "name")
+		else:
+			associated_doc = frappe.db.get_value(
+				"File",
+				{"content_hash": self.content_hash, "name": ["!=", self.name], "is_folder": False} # type: ignore
+			)
+
+		if associated_doc:
+			existing_file = frappe.get_doc("File", associated_doc)
+			existing_file.attached_to_doctype = self.attached_to_doctype
+			existing_file.attached_to_name = self.attached_to_name
+			self.content_hash = existing_file.content_hash
+			self = existing_file
+
+		existing_attachment = list(
+			filter(
+				lambda row: row.link_doctype == self.attached_to_doctype
+				and row.link_name == self.attached_to_name,
+				self.file_association,
+			)
+		)
+		if not existing_attachment:
+			self.append(
+				"file_association",
+				{"link_doctype": self.attached_to_doctype, "link_name": self.attached_to_name},
+			)
+		if associated_doc:
+			self.save()
+
 	def validate(self) -> None:
+		self.associate_files()
 		if self.flags.cloud_storage or self.flags.ignore_file_validate:
 			return
 		if not self.is_remote_file:
 			super().validate()
 		else:
 			self.validate_file_url()
 
+	def after_insert(self) -> File:
+		if self.attached_to_doctype and self.attached_to_name and not self.file_association:
+			if not self.content_hash and "/api/method/retrieve" in self.file_url:
+				associated_doc = frappe.get_value("File", {"file_url": self.file_url}, "name")
+			else:
+				associated_doc = frappe.db.get_value(
+					"File", {"content_hash": self.content_hash, "name": ["!=", self.name], "is_folder": False}
+				)
+			rename_doc(self.doctype, self.name, associated_doc, merge=True, force=True)
+
+	def remove_file_association(self, dt: str, dn: str) -> None:
+		if len(self.file_association) <= 1:
+			self.delete()
+			return
+		for row in self.file_association:
+			if row.link_doctype == dt and row.link_name == dn:
+				frappe.delete_doc("File Association", row.name)
+				del row
```

Comment:
I think you can just do:

```suggestion
				self.remove(row)
```

Reply:
The docs say this should work, but it doesn't. `del` is more reliable in that you don't need the list's permission and matching function to remove it. Also more performant.